### DISCUSSION TITLE
v2.0.0 Beta

### DISCUSSION THREAD

**Discussion by jacobbednarz** [cloudflare/cloudflare-go#1538] [Announcements]
> [!WARNING]
> 
> `v2.x` of this library is a ground-up rewrite of the SDK, using code generation from our OpenAPI spec.
>
> This package should be in a usable state for many projects, but expect frequent minor breaking
> changes as we rename methods and types until GA. If this isn't suitable for you project, we recommend
> pinning to the `0.x` releases for now.

We're working on a new major version of our SDK and are excited to have you test it!

Please use this discussion page as a place to report issues and feedback about the new release.

## Install

```
go get -u github.com/cloudflare/cloudflare-go/v2
```

## Code

https://github.com/cloudflare/cloudflare-go/tree/v2

## Things we would like feedback on

- Endpoints that do not work despite existing in the SDK
- Types and error handling
- Major regressions from 0.x releases

## Migration guide

[0.x to 2.x upgrade guide](https://github.com/cloudflare/cloudflare-go/discussions/1537)

**Comment by Geun-Oh** [comment]
Hello. I’m a user of cloudflare kv package with go.
I have 2 little suggestions.

\# 1

In the process of listing and deleting keys in the namespace, I went through the process of checking `ListWorkersKVsParams.Cursor` to know if there is remain keys that I didn’t get.  

This is an example of it.  (ver 0.9x)
```go
for {
	resp, err := s.api.ListWorkersKVKeys(context.Background(), cloudflare.AccountIdentifier(s.accountID), cloudflare.ListWorkersKVsParams{
		NamespaceID: s.namespaceID,
		Cursor:      cursor,
	})

	if err != nil {
		log.Printf("Error occur in ListWorkersKVKeys: %v", err)
		return err
	}

	keys = make([]string, len(resp.Result))

	for _, element := range resp.Result {
		name := element.Name
		keys = append(keys, name)
	}

	_, err = s.api.DeleteWorkersKVEntries(context.Background(), cloudflare.AccountIdentifier(s.accountID), cloudflare.DeleteWorkersKVEntriesParams{
		NamespaceID: s.namespaceID,
		Keys:        keys,
	})

	if err != nil {
		log.Printf("Error occur in DeleteWorker: %v", err)
		return err
	}

	if len(resp.Cursor) == 0 {
		log.Println("No keys left in the namespace")
		break
	}

	cursor = resp.Cursor
}
```

But in v2, I found a `GetNextPage()` method at pagination internal package. But in v2, I found a GetNextPage() method at pagination internal package. Though it’s a good news to me, as the way of listing all keys has quite difference compared with latest version, let me know that do you have mind to add a migration guide or example.

\# 2

Would you add a flag of `list_complete`(like `Done` flag before!)? We can see it on [Cloudflare API Docs with JS](https://developers.cloudflare.com/kv/api/list-keys/) and it would be great to implemented in here.
Maybe it can improve DX much better.

If there's something that I'm wrong about, please tell me. :)

**Reply by jacobbednarz** [reply]
👋 thanks for the feedback.

as of today, the migration we offer is through the GritQL pattern linked and the manual notes above. we're not able to do a method by method (or even product) deep dive largely because there isn't a 1:1 for all products or methods. at this point, we won't be adding any further migrates to the v0 => v2 plan. for future major versions, we'll be looking to provide more comprehensive auto migrations.

for your second point on knowing when the pagination finishes, have you tried the [guidance in the README](https://github.com/cloudflare/cloudflare-go/tree/v2?tab=readme-ov-file#pagination)? with `.Next()` you shouldn't need a `Done` or `HasMorePages` method. let me know how you go.

**Reply by Geun-Oh** [reply]
Thanks for your reply. I understand the things about migration.

And, thanks for noticing me about `.Next()`.  I missed it before 😂.

**Comment by acterry** [comment]
What's the future of flarectl once v2 ships? I was just lamenting that the current flarectl doesn't support the rules engine.

A full-featured CLI would take my Cloudflare config experience to Cloud 9.

I don't see that Stainless supports generating a CLI, but I see there are other tools that claim to be able to do so.

Is Cloudflare planning to offer a CLI that's autogenerated from the OpenAPI Spec -- or will it be left to the community to provide something?

**Reply by jacobbednarz** [reply]
there will be a CLI in the future. it won't be `flarectl` though and we're not extending the current `flarectl` any further. no ETA to share yet but watch this space.

**Reply by acterry** [reply]
@jacobbednarz , if there's ever an opportunity for customers (enterprise or otherwise) to beta test, count me in.

**Reply by jacobbednarz** [reply]
feel free to shoot me an email `jb@` with your account ID and use case and i'll add a note for when it comes up. 

betas are generally announced on our socials or the repositories so if you're following there, you should see it pop up.

**Comment by cylussec** [comment]
It would be great to have mocks as a part of the new library. I am incorporating unit tests with my code, and it would be great to be able to just import existing ones.

**Reply by jacobbednarz** [reply]
we don't have plans to export mocks from the library at this point but will note it down for future investigation.  a local testing solution is under way, but for now, a mix of prism and our OpenAPI spec meets most needs (when prism works 🙄)

**Comment by favonia** [comment]
Hi, I'm trying out the v2 library and in general I'm happy with a group-up clean-up. Unfortunately I noticed an issue that prevents me from adopting it and I hope we can resolve it. Currently, the `NewClient` function will automatically look up various environment variables and collect them as options. While this is convenient for implementing a specific application that uses these variables in the exactly same way, I feel cloudflare-go as a library probably should not do this. It can perhaps provide helper functions to generate those options, but there should be a clean `NewClient` function that only takes in options given as function arguments.

**Reply by jacobbednarz** [reply]
thanks for kicking the tyres here @favonia. 

regarding pulling in the default credentials from the environment variables, that's something we have been asked for as part of a zero configuration client when used with existing Cloudflare tooling. 

you said you don't think it should do this; can you elaborate on why?

**Reply by favonia** [reply]
To be clear, I am not against pulling in credentials from the environment variables, but the lack of a "clean" client when the library is used as part of another application. I believe there are several drawbacks in building in such pulling (and providing no ways to disable it):

1. Testing becomes more challenging because it directly calls `os.LookupEnv`.
2. While my application also uses environment variables, it is currently following the older `CF_*` convention. Handling the deprecated and new variables would render the built-in pulling redundant. Moreover, I specifically do not want to support `API_KEY` and the library is forcing me to work around the automatic pulling to block it. Additionally, my application strips off whitespace, warns about common mistakes, and accept other options (e.g. reading from files). The automatic pulling makes implementing these extra features more challenging.
3. I know many other applications (not mine) would prefer pulling the credentials from JSON files or other places. Many of them are probably not using cloudflare-go but I would use them as examples: it's not obvious to me that one has to override or block the default pulling when upgrading the library. I was able to discover this issue only because I read a lot of documentation. This kind of bugs would be really difficult to discover because everything will appear to work fine.
4. Some Cloudflare tools also read `CLOUDFLARE_API_BASE_URL` while the current library does not. I am not suggesting that this library should also read `CLOUDFLARE_API_BASE_URL`. Rather, my point is that it is not conceptually clear which environment variables will be read and which will not. The API thus adds some mental burden on the programmers, especially when the pulling cannot be disabled. That is, if I put this aside and come back some time later, I don't think I can remember the exact four environment variables this library will automatically read and will have difficulty seeing why my application (with all the workarounds) has handled all the variables correctly. Such hidden dependencies are difficult to remember. If there were a function to create a "clean" client, it would be much easier to confirm that some code is doing the right thing.
5. It is easy enough to rename `NewClient` to `NewClientWithDefaultsFromEnvironmentVariables` (or something shorter and less scary). My preferred approach would be a function `DefaultsFromEnvironmentVariables` such that `NewClient(FromEnvironmentVariables()...)` recovers the current behavior. At very least there should be `NewCleanClient` with absolutely no default options pulled from the environment.

In any case, I think there should an easy way to create a "clean" client, and I hope that could be the behavior of `NewClient`.

**Reply by jacobbednarz** [reply]
by default, we intentionally don't want a "clean" client. we are optimising for a zero configuration adoption that will co-exist alongside other tools. we may look at adding another method to do it, however, it's not in the immediate roadmap. other options are also floating around how we handle "blocking" of environment variable blocking which may change this here.

> Some Cloudflare tools also read CLOUDFLARE_API_BASE_URL while the current library does not.

we intentionally have moved it to a client instantion option instead (like HTTP clients) as it is more idiomatic and not sensitive like the other values so less likely to be an environment variable. you can see the docs at: https://github.com/cloudflare/cloudflare-go/blob/v2/option/requestoption.go#L26-L35 if you want to continue using it.

> Moreover, I specifically do not want to support API_KEY and the library is forcing me to work around the automatic pulling to block it. 

i can't comment on your reasoning here, however, this approach won't work for all endpoints. while API token coverage is pretty good, there are some things that can only be performed with API key/email combo today and not supporting it does limit what you can implement.

**Reply by favonia** [reply]
@jacobbednarz Thanks. I think the fundamental difference in our thinking is that I view `cloudflare-go` as a generic and flexible library, possibly with helper functions facilitating implementations of Cloudflare's in-house tools. This means that the library is _not_ only for implementing Cloudflare's in-house tools, and that's why I feel a bit uncomfortable with the current design. I feel there are a few things Cloudflare should consider regarding the zero configuration:

1. Cloudflare can officially endorse the universal "zero configuration" environment variables `CLOUDFLARE_*`, force all in-house tools to follow the same specification, and encourage all third-party tool developers to follow suit. Once Cloudflare officially endorses the scheme, it's natural and expected that the `NewClient` will read those environment variables. This will place those special environment variables at the same level of `TZ` and `HTTP_PROXY` within the Cloudflare community.
2. However, while standardizing this behavior, `cloudflare-{go,...}` libraries should help applications transition from their existing behaviors. This is what's severely lacking in the current `v2` code and why I want a "clean" client to implement my own transitional handling.

An inspiring example is the treatment of `HTTP_PROXY` in the Go standard library. It has the following definition as `DefaultTransport`:
```go
var DefaultTransport RoundTripper = &Transport{
	Proxy: ProxyFromEnvironment,
	DialContext: defaultTransportDialContext(&net.Dialer{
		Timeout:   30 * time.Second,
		KeepAlive: 30 * time.Second,
	}),
	ForceAttemptHTTP2:     true,
	MaxIdleConns:          100,
	IdleConnTimeout:       90 * time.Second,
	TLSHandshakeTimeout:   10 * time.Second,
	ExpectContinueTimeout: 1 * time.Second,
}
```
where [`ProxyFromEnvironment`](https://pkg.go.dev/net/http#ProxyFromEnvironment) is what reads the standard environment variables. This gives developers the ability to turn off the reading of those environment variables while having sensible default behaviors. The Go design is that the zero value of `http.Client` uses the above `DefaultTransport`. I'm not sure how exactly to adopt this design in the current `v2` codebase, but as I suggested earlier, we can have `cloudflare.NewClient(option.ZeroConfig)`, `cloudflare.NewZeroClient()` and/or `cloudflare.NewCleanClient()`. I see many possibilities to keep it easy for the zero configuration while making the library flexible for non-Cloudflare developers, and more importantly all these changes only require simple refactoring of the current code.

As for API keys, thanks for the information that not all end points are covered by API tokens yet. (What are they? I'm curious.) That said, my application only needs a handful of endpoints that _are_ covered by API tokens and I still believe it's more future-proof to use API tokens. (Please correct me if I'm wrong.) If my understanding is correct, I do hope there's a way to at least warn about the use of API keys.

**Reply by favonia** [reply]
@jacobbednarz Updates: I just released a new version of my tool that will also accept variables `CLOUDFLARE_*`, making it compatible with the new Cloudflare convention. Also, just to be clear, I'm still very interested in adopting the new cloudflare-go library, especially if there's a slightly easier way to set up a clean client. It has so many other improvements I wish to have. I'm fine with very long function names such as `NewUnconfiguredClient`.

**Reply by jacobbednarz** [reply]
thanks for the updates. at the moment, the "clean" or explicit client you're after is still being discussed and working out what that looks like across the board.

regarding API tokens, you can check out the new docs at https://developers.cloudflare.com/api-next/ for the endpoints and it will include the available auth schemes.

**Reply by favonia** [reply]
@jacobbednarz Thanks for the information. The new site looks great! I will take a more careful look later.

**Comment by favonia** [comment]
**This has been fixed**

---

I have another question: apparently I don't understand how `param` works. Why doesn't the following field have the type `param.Field[string]`?
```go
type AccountGetParams struct {
	AccountID param.Field[interface{}] `path:"account_id,required"`
}
```

**Reply by jacobbednarz** [reply]
`param.Field` is a wrapper type to allow determining the full state of a value (such as is set, is known, etc)

the `interface{}` here is what happens when schemas are under defined and lacking a concrete type in the openapi schema definition. teams have tickets to address these and flow through automatically as they do.

**Reply by favonia** [reply]
I see. I will wait for the correction, then.

**Reply by favonia** [reply]
This has been corrected. (Is there a way to mark this as "done"?)

**Comment by favonia** [comment]
Regarding error handling, the 2.x approach looks much better than 0.x, but I don't understand this principle about avoiding wrapping:

> When other [non-API] errors occur, they are returned unwrapped

Why should we avoid wrapping in other cases? `errors.Is` and `errors.As` can handle wrapping fine. BTW, I wonder if we could reduce the use of pointers---I heard that pointers help only when copying a relatively large structure.

**Comment by pawelhenek** [comment]
Hello, 
I just wanted to ask if the is any reason why `Deploy hooks` are not documented in Cloudflare API docs and thus there is no cloudflare-go and cloudflare terraform provider support? I mean I'm considering contributing to the library (which version should I start with main or v2?) and then integrate it to terraform provider, but wondering why it's not even documented. Shouldn't be considered as stable so contributions are not good idea? 
The endpoint I meant  
`https://dash.cloudflare.com/api/v4/accounts/<ACCOUNT_ID>/pages/projects/<PROJECT_NAME>/deploy_hooks
`
Thanks in advance
